File: src/com/OpenSerenity/elements/BaseListElement.java
package com.OpenSerenity.elements;

import java.util.List;

public abstract class BaseListElement<TElement extends BaseElement> extends BaseElement<TElement> {
    public abstract void selectByIndex(int index) throws Exception;
    public abstract void selectByText(String optionText) throws Exception;
    public abstract String getSelectedItemText() throws Exception;
    public abstract List<TElement> getItems() throws Exception;
    public TElement get(String text) throws Exception {
        for(TElement element : getItems()) {
            if(text.equals(element.getText())) {
                return element;
            }
        }
        throw new Exception("element not found");
    };
    public TElement get(int index) throws Exception {
        return getItems().get(index);
    };
}

File: src/scala/com/OpenSerenity/elements/BaseListElement.scala
package com.OpenSerenity.elements

abstract class BaseListElement[TElement <: BaseElement[TElement]] extends BaseElement[TElement] {
  def selectByIndex(index: Int)
  def selectByText(optionText: String)
  def getSelectedItemText(): String
  def items(): List[TElement]

  def get(text: String): TElement = items find(text == _.getText()) get
  def get(index: Int): TElement = items()(index)
}